### DISCUSSION TITLE
Connection.execute is very slow to run.

### DISCUSSION THREAD

**Discussion by connorslade** [rusqlite/rusqlite#1119] [Questions] [answered]
I'm new to rusqlite and SQL.

I'm trying to increment the value of a cell, but when running Connection.execute to update the value, it is very slow.
It is averaging `~30ms` for each execution, and this is much higher than I would expect.

Here is a section of the code I am using, with the time for each bit to execute.
```rust
let conn = rusqlite::Connection::open("data/data.db").unwrap(); // 108us

let views: u32 = conn
    .query_row(
        "SELECT views from article where name = ? LIMIT 1",
        [doc.path.clone()],
        |row| row.get(0),
    )
    .unwrap(); // 185us

let views = views + 1;

conn.execute(
    "UPDATE article SET views = ?2 WHERE name = ?1",
    rusqlite::params![doc.path.clone(), views],
)
.unwrap(); // **42ms**

conn.close().unwrap(); // 113us
```

Any help would be appreciated.

**Comment by gwenn** [comment]
Why not using a single query ?
```sql
UPDATE article SET views = views + 1 WHERE name = ?1;
```

**Reply by connorslade** [reply]
I split it into two queries because I need to access `views` later in the code.

**Reply by gwenn** [reply]
> I split it into two queries because I need to access `views` later in the code.

https://sqlite.org/releaselog/3_35_0.html and [https://sqlite.org/lang_returning.html](RETURNING)
```sql
UPDATE article SET views = views + 1 WHERE name = ?1 RETURNING views;
```

**Comment by thomcc** [comment, marked as answer]
Are you building with `--release` (e.g. `cargo build --release` or `cargo run --release`)? Those numbers are unreasonably slow for sure.

If you're using `features = ["bundled"]`, this controls the optimization level of SQLite too. (And if you aren't, it might be worth considering, as it's likely more up to date than the system version, and generally newer SQLite versions are faster than older ones, as they have more optimizations)

That said, this could also indicate insufficient indices in the db -- are the timings similar if you use the CLI tool? `sqlite3 data/data.db` and then run `.timer on` before you do your queries

**Reply by thomcc** [reply]
Looking at the queries, you probably want some index on `articles(name)` if you don't have one. An easy way to do this might be to execute a `CREATE INDEX IF NOT EXISTS articles_name_index ON articles(name);` after connecting.

(Also, if `name` is unique across all rows of `articles`, a `CREATE UNIQUE INDEX` would be better, but not that this would make attempts to insert a new row in articles if its name is the same as some other row)

Without something like this, it means SQLite has to search every row of the table in order to find the name, which is a lot of work.

**Reply by connorslade** [reply]
I am building with release and using bundled SQLite. With the SQLite shell, I am getting the same times.

I tried your suggestion of using an index, but that didn't help either. This database only has 11 rows.

**Reply by thomcc** [reply]
If the SQLite shell is slow too, then it probably has something to do with I/O performance on your machine, unfortunately. I also misread us as ms before -- 40ms is surprisingly slow for writes, but not impossibly slow.

There are some things you can try here (the biggest one might be executing `PRAGMA journal_mode=WAL` on startup to enable WAL mode: https://sqlite.org/wal.html), but it's hard to give general advice here without knowing more.

Also, note that it's generally faster to perform things inside a transaction. This won't make a difference for single queries, but if you have to insert 1000 rows, then definitely do that in a transaction, which will avoid many of the costs until the final commit.

(I'd also recommend a transaction here for correctness reasons around the select-then-insert, but its neither here nor there, and the update is better)

**Reply by connorslade** [reply]
Setting the `journal_mode` to `WAL` reduced the execution time to `~6ms` which is more acceptable.
I was also experimenting with other pragma statements and found that setting `synchronous` off reduced the time to `~20us`.
So I guess most of the delay was just waiting on the filesystem after every update.

**Reply by thomcc** [reply]
Ah... I'd be careful there `PRAGMA synchronous=off` outside of tests or databases with contents you don't care about. For example, any power loss might probably corrupt data -- hard shutdowns[^1], laptop/phone battery dying, accidentally unplugging it, etc. IMO it's not worth it really (one of the best things about SQLite is being able to avoid issues here..), but it's your call.

[^1]: Like, when you hold down the power button till the computer is choked out.

That said, for WAL there should be no real downside for `PRAGMA synchronous=NORMAL`, which should be faster than the default -- I had forgotten that you should set that too, which has very few downsides under WAL mode.

If you need maximal performance, then another thing that might help is using the `unix-excl` VFS when on unix (there's no equivalent for windows). This is faster, but doesn't support multiple processes[^2] using the DB simultaneously -- the 2nd process to try to open will get an error. You can do this like:

```rust
if cfg!(unix) {
    Connection::open_with_flags_and_vfs(path, rusqlite::OpenFlags::default(), "unix-excl")
} else {
    Connection::open(path)
}
```

Other performance things are too hard to judge with a DB that small. For example, changing the cache sizes and page sizes could also help, but for such a small DB, the numbers may be different -- for example, larger page size will definitely hurt if the DB is tiny, but it can be beneficial for larger ones. Actually, this is true always; it's pretty hard to measure this stuff for such a small db, and especially a DB that's only just been created. The first queries you perform after opening a connection will never hit any of SQLite's caches and stuff like that.

[^2]: It still supports multiple connections in the same process though, even across multiple threads.

**Reply by connorslade** [reply]
`synchronous=NORMAL` works great! The execution time is `~10us` now. VFS didn't reduce the time any more, but it's already more than acceptable. Thank you for your help.

I guess I will mark this thread as answered now.
